parser.go

```
@@ -0,0 +1,191 @@
+			}
+			if !closingParaFound {
+				ErrSectionNameMissingClosure = fmt.Errorf("section [%s] is missing closure paranthesis ']'", line[1:])
+				return ErrSectionNameMissingClosure
+			}
+			if parsedMap.dictionary == nil {
+				parsedMap.dictionary = make(map[string]map[string]string)
+			}
+
+		} else if line[0] == ']' {
+			ErrWrongParanthesisOrder = fmt.Errorf("wrong section paranthesis order, the section provided, %s, cannot start by ]", line)
+			return ErrWrongParanthesisOrder
+		} else if line[0] == ';' || line[0] == '#' || line[0] == ' ' || line[0] == '\n' || line[0] == '\t' {
+			continue
+		} else {
+			equalFound := false
+			for j, ch := range line {
+				if ch == '=' {
+					equalFound = true
+					key = line[0:j]
+					value = line[j+1:]
+					key = strings.Trim(key, " ")
+					value = strings.Trim(value, " ")
+					parsedMap.dictionary[section][key] = value
+					break
+				} else if ch == ']' {
+					ErrInvalidSectionName = fmt.Errorf("section name provided, %s, can't start with anything other than '['", line[:j+1])
+					return ErrInvalidSectionName
+				}
+			}
+			if !equalFound {
+				ErrMissingValueAssignment = fmt.Errorf("key provided, %s , is not assigned to a value, no '=' found", line)
+				return ErrMissingValueAssignment
+			}
+		}
+	}
+	return nil
+}
+
+// GetSectionNames provides section names of parsed file/string
+func (parsedMap *Parser) GetSectionNames() []string {
+	var sectionNames []string
+	for section := range parsedMap.dictionary {
+		sectionNames = append(sectionNames, section)
+	}
+	return sectionNames
+}
+
+// GetSections provides the dictionary/map structure
+func (parsedMap *Parser) GetSections() map[string]map[string]string {
+	return parsedMap.dictionary
+}
+
+// Get function takes 2 parameters: section and its key
+// Provides equivalent value
+// If section or key aren't found, it returns the zero values
+func (parsedMap *Parser) Get(sectionName, key string) string {
+	return parsedMap.dictionary[sectionName][key]
```

if the section doesn't exist this would panic

---

> if the section doesn't exist this would panic

But the thing is, even if section doesn't exist map of golang by default returns zero value of everything
do we need to return an error here? I can create another err

---

yes you are right, tried it now, writing is the one that panics

---

nope you shoudn't return any errors